src/control/tools/EditSelectionContents.cpp

```diff
@@ -381,9 +381,8 @@ void EditSelectionContents::finalizeSelection(double x, double y, double width,
 
 	bool move = mx != 0 || my != 0;
 
-	auto iwl = this->indexWithinLayer.rbegin();
-	auto e = this->selected.rbegin();
-	for (; iwl != this->indexWithinLayer.rend() and e != this->selected.rend(); ++iwl, ++e)
+
+	for (auto e = this->selected.rbegin(); e != this->selected.rend(); ++e)
```

Comment:
Maybe this is we would get a whiteboard and hash it out if we were co-located. I'll try to explain and we'll see what's up. I'm not sure if the reversing seems weird or something else.

Regarding the reversing:
Suppose we had three elements in the layer:
[bottom] A - B - C [top]
Suppose both B and C are selected, and assume the selections are done sequetially with B first, then C second.
First, B is added to the selection vector [B], its index is calculated (which is 1) and it's put into the map (B -> 1). The layer now looks like this:
[bottom] A - C [top]
C is now added to the selection vector [B, C], its index is calculated (which is 1 again) and it's put into the map (C -> 1, B->1).

The items need to be returned to the layer in reverse order to how they were given - e.g. C first rather than B. That's why the iterator is reversed. 

Regarding adding elements with an InvalidIndex, I think we're fine. If at any point there is a pasted item added on top (as is the default with an Invalid Index) it is impossible for any item to be inserted on top of that.

Reply:
Worried I missed your point here, let me know :)

Reply:
Right, so `selection` essentially becomes a stack.

Reply:
The problem I meant is, that the selection vector is also used to draw the selection.
So if {A:InvalidIndex, B:InvalidIndex, C:InvalidIndex} is in the vector, it will be drawn A->B->C.
But after releasing the selection it will be reverted since we add it in reverse order.

My way to fix it temporarily would be:
*Add a `std::deque insert_order;` next to the `std::vector selection`.
*use `std::deque::emplace_front` to add all Elements with a valid Index.
*use `std::deque::emplace_back` to add all Elements with an invalid Index.

You can test the current behaviour by drawing some rectangles, cut them out, and add them back. After releasing the selection, the order will be inverted.
Got some screenshots:
![grafik](https://user-images.githubusercontent.com/2228627/71752482-837db780-2e7f-11ea-99dd-a61c6fad414d.png)
![grafik](https://user-images.githubusercontent.com/2228627/71752489-8aa4c580-2e7f-11ea-9bf1-c5e4193a5c30.png)

(PS: I saw, that you want to refactor the part of the code anyway, so the change is optional, but please do it, when you have a limited time and you don't know when you can refactor the part)